MNT Enforce ruff rules (RUF)

DimitriPapadopoulos opened the pull request:
#### What does this implement/fix? Explain your changes.

Enforce [Ruff-specific rules (RUF)](https://docs.astral.sh/ruff/rules/#ruff-specific-rules-ruf) rules.

#### Any other comments?

Some rules have been disabled: either they don't seem that interesting, or they change too much of the codebase.

[doc skip]

DimitriPapadopoulos committed:
MNT Address ruff warning

warning: Invalid `# noqa` directive: expected `:` followed by a comma-separated list of codes (e.g., `# noqa: F401, F841`).

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF010

RUF010 Use explicit conversion flag

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF017

RUF017 Avoid quadratic list summation

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF022

RUF022 `__all__` is not sorted

DimitriPapadopoulos committed:
MNT Enforce ruff rules (RUF)

jeremiedbb commented:
Thanks for the PR @DimitriPapadopoulos. These rules mostly summarize to ordering `__all__` and properly deal with `str` and `repr` in fstrings. I'm okay with that. Some comments though

- I checked enabling `RUF021` and there was no error. Maybe we can remove it from the exceptions ? This rule is "Parenthesize a and b expressions when chaining and and or together, to make the precedence clear" which I think is fine to enforce.

- `RUF015` seems good to have imo. I always find these `list(some_generator)[0]` a bit clunky. I checked enabling it and there are not many occurrences to fix so maybe worth to remove it from the exceptions ?

- How come we can remove all these `noqa` ?

jeremiedbb commented on sklearn/neighbors/_base.py:
```diff
@@ -1249,13 +1250,13 @@ class from an array representing our data set and ask who's
-                neigh_dist_list = sum(neigh_dist_chunks, [])
-                neigh_ind_list = sum(neigh_ind_chunks, [])
+                neigh_dist_list = reduce(operator.iadd, neigh_dist_chunks, [])
+                neigh_ind_list = reduce(operator.iadd, neigh_ind_chunks, [])
                 neigh_dist = _to_object_array(neigh_dist_list)
                 neigh_ind = _to_object_array(neigh_ind_list)
                 results = neigh_dist, neigh_ind
             else:
-                neigh_ind_list = sum(chunked_results, [])
+                neigh_ind_list = reduce(operator.iadd, chunked_results, [])
```
I don't find these ones much better. And I don't think it's the recommended way to do that. I would be okay to replace with
`list(itertools.chain.from_iterable(neigh_dist_chunks))`
What do you think ?

DimitriPapadopoulos replied:
Sure. That's probably the most readable of all, at least to me.

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF015

RUF015 Prefer `next(...)` over single element slice

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF021

RUF021 Parenthesize `a and b` expressions when chaining
       `and` and `or` together, to make the precedence clear

lesteve commented:
@DimitriPapadopoulos I think you have contributed [useful stuff in the past to the project](https://github.com/scikit-learn/scikit-learn/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Adimitripapadopoulos), thanks a lot :pray:!

In this case it looks like you opened multiple PRs on separate aspects on cleaning up/maintenance. Sometimes splitting PRs is good, and sometimes as a busy maintainer it can feel a bit too much.

I would suggest you keep only one PR open and close the others, the one you think is the most important, wait for feed-back, iterate until it is merged, and then open another PR ...

My personal opinion on this kind of PRs, is that they don't have huge benefits and they tend to be controversial. People will have different opinions on whether it is worth it or not, and at the end significant reviewer and contributor time has been spent on something that probably was not so crucial for the project and this time could have been invested somewhere else with a more useful outcome ...

DimitriPapadopoulos commented:
@lesteve Other maintainers insist on smaller PRs - tastes and colors.

While it's possible to discuss individual rules, linters do result in more maintainable code overall, and sometimes find actual bugs. Perhaps the benefit is not huge, but the cost is small. As for the controversial part of rules, choosing a mainstream linter (or formatter) gets controversy out of the way.

As for closing the PRs, now that they're already opened, I don't know.

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF100

RUF100 Unused blanket `noqa` directive

DimitriPapadopoulos committed:
MNT Apply ruff rule RUF002

RUF002 Docstring contains ambiguous ` ` (NO-BREAK SPACE). Did you mean ` ` (SPACE)?

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -341,7 +341,7 @@ def euclidean_distances(
     Notes
     -----
-    To achieve a better accuracy, `X_norm_squared` and `Y_norm_squared` may be
+    To achieve a better accuracy, `X_norm_squared` and `Y_norm_squared` may be
```
I can not tell what this change is. A unicode character?

DimitriPapadopoulos replied:
There still is a single commit per rule, so there's the workaround of reviewing the commits individually.

In this case, I manually changed a NO-BREAK SPACE U+00A0 into a regular SPACE U+0020 in https://github.com/scikit-learn/scikit-learn/pull/30694/commits/0aec43b51a914e237683a09e0a569d9e967cb903, to silence [ambiguous-unicode-character-docstring (RUF002)](https://docs.astral.sh/ruff/rules/ambiguous-unicode-character-docstring/#ambiguous-unicode-character-docstring-ruf002).

thomasjpfan reviewed (approved):
For a reviewer, it is a lot of different types of changes to go through. I think if we want to enforce more linting rules, it's best to introduce them one at a time. i.e. One small PR on sorting `__all__`, one on removing unncessary `# noqa`, etc and slowly introduce them one at a time.

In any case, given we already have this PR, I'm okay with enforcing the proposed rules.

adrinjalali reviewed (approved):
LGTM. You need to add a this to `.git-blame-ignore-revs` after merging this.

DimitriPapadopoulos commented:
> LGTM. You need to add a this to `.git-blame-ignore-revs` after merging this.

Do you expect me to add teh commits to `.git-blame-ignore-revs` in an upcoming PR? I could do that for all recent ruff-related PRs.

Note however that these are most commits are automatic ruff fixes, but I manually fixed some issues.

adrinjalali commented:
Yes, we prefer to completely ignore styling changes to keep the blame relevant in case we need to know who worked on a feature.
